src/OrleansRuntime/GrainTypeManager/GrainTypeData.cs

```diff
@@ -151,5 +154,11 @@ internal static PlacementStrategy GetPlacementStrategy(Type grainClass, Placemen
 
             return predicate;
         }
+
+        private ParameterInfo[] GetConstructorParameters()
+        {
+            var constructor = this.Type.GetConstructors(BindingFlags.Public | BindingFlags.Instance).FirstOrDefault();
+            return constructor?.GetParameters() ?? EmptyParameters;
+        }
```

Comment:
> If users want to deviate from this, we can't stop them, but this is the only pattern we should support.

I agree with it. However, we should add this to the (long) list of validations (like always return `Task/Task<T>`, don't use `lock`, etc.)  that we _should_ provide in an optional Roslyn analyzer package.